old_modules/m_aux3.man
.SH "A/UX NOTES"
There is no obvious way to find the amount of memory per process
that is currently resident.  The percent cpu usage is not
maintained by the operating system, so we fake one.
The last pid number is not plucked from the operating system
either.  The number given is based on the information collected
in the process snapshots.
.LP
Although AUX does not generally have a renice system call,
there is a compile-time option that may have been enabled
when the program was installed.
.LP
The A/UX 3.1 port was written by Richard Henderson <rth@tamu.edu>.


Changes
Wed Apr 18 2007 - wnl (3.6.1)
	Fixed a few bugs in sigconv.awk that were causing incorrect
	results on FreeBSD.  Changed configure.ac to fix a few linux
	problems: signal include file and /proc/1/stat.

Fri Apr 13 2007 - wnl (3.6.1)
	Removed the use of VPATH for compiling the system module and used
	an explicit dependency in the Makefile instead.  VPATH is now set
	to just srcdir to ensure that top will compile correctly when
	configured from a different directory.  On systems without VPATH
	support, top will still configure and compile, but only
	from within the source directory.  This fixes bug 1699526.

Fri Feb  2 2007 - wnl (3.6.1)
	Revised the way that configure figures out owner, group, and mode.
	For systems that don't use the kernel, it tries to match install
	settings to allow access to stuff in /proc.  More importantly, if
	mode is 755 then neither owner nor group are set.  This fixes bug
	1631136.  Added patch from haanjdj@xs4all.nl to fix an occasional
	core dump in m_decosf1.c.  This checks return code from task_threads.
	Made sure all get_system_info functions are declared void.  Fixed
	string termination bug.  Cleaned up documetation for sunos5.

Tue Aug  8 2006 - wnl (3.6.1)
	For Solaris, changed the tag "swap" to "total swap" to clarify
	what is beign displayed.  Note that the calculations are still the
	same: the display is just showing total rather than total - free.

Thu Apr 27 2006 - wnl (3.6)
	Added patches for linux-style sort shortcuts and for Unixware
	support in configure (patch 1474427).  Fixed sunos5 to do slow start
	and to ensure cpucount is set (patch 1477386).  Added pagination
	routines to display.c and modified show_help to use it, since the
	help screen is now longer than 24 lines.  Applied patch for unixware
	support that adds check for mas library (patch #1474423).  Solaris
	cpu percent now reflects a percentage of the entire server, rather
	than a single cpu (bug 1478138).

Mon Mar 27 2006 - wnl (3.6)
	The production release of version 3.6.  Fixed a minor scaling
	bug in the decosf1 module.  Support for MacOS X is officially
	withdrawn although the macosx module is still part of the
	distribution.  Hopefully this is a temporary situation.
	Documentation updated.


Wed Feb 15 2006 - wnl (3.6beta5)
	Minor changes to eliminate warnings from the Sun Studio compiler.
	These were mostly sloppy argument declarations.  I also added
	message.h to provide an interface file for just the message
	related functions of display.c.

Mon Dec 26 2005 - wnl (3.6beta4)
	Added new netbsd module, courtesy of Simon Burge.
	Fixed a few bugs in SVR4 module and added its use to
	configure.ac, thanks to Sanchet Dighe.  Also ensured that the
	novpath Makefile was in the distribution.
	Fixed portability problem in display.c


Mon Oct 24 2005 - wnl (3.6beta3)
	Set up a color tagging mechanism in color.c to allow for the
	dynamic creation of tag names to contol color highlighting.
	These names are partially derived from the tags used to label
	memory and swap information on the screen, thus are driven by
	the machine module itself.  Added -T option to list color 
	highlighting information.  Help screen now includes the actual
	list of sort order names.  Incorporated some minor fixes to
	the main code from the Freebsd source tree.  Fixed bug #1324582.
	Freebsd 5: removed WCPU column and added THR column.  Display
	for freebsd 4 and earlier unchanged since they don't track
	threads in the kernel.  Added LICENSE file to distribution.

Wed Oct 12 2005 - wnl (3.6beta2)
	Major overhaul to display.c.  All lines of the display are
	directly tracked and controlled via display_write and its
	companion display_fmt.  Added support for complete control
	of ANSI color on the screen: this will be used in the future
	to allow for full use of color everywhere on the screen.
	Signal handling code now uses sigaction on all systems that
	support it.  Restored the freebsd module and did away with
	freebsd4, and upgraded freebsd module to support 5.x.
	Fix bug #1306099 (wio(wait) timer ignored on OSF1).

Fri Sep 23 2005 - wnl (3.6beta1)
	Fixed bugs #1266341 (compilation errors with gcc 4.x),
	#1156464 (cpu% field for sunos), #1156243 (compilation
	errors on AIX).  Applied patches #1217855 (Solaris 10
	xarch flag).  Overhaul of sunos5 module, making code more
	efficient and easier to follow.  Got rid of need for MEMTYPE
	type in utils.h.  Changed all memory statistics data in the
	module specification from an int to a long to better support
	64-bit systems with lots of memory. Moved all unused modules
	out of the distribution (I will add them back in as needed).
	Moved freebsd module to freebsd4 as it won't work with 5.x
	(a new module will be necessary). Added support to configure
	for makes that don't understand VPATH. Updated documentation:
	man page, FAQ, README, INSTALL.

Mon Jan 24 2005 - wnl (3.6alpha10)
	Updated aix43 module with ANSI function declarations and fixed
	declaration of get_system_info.  Configure now uses irixsgi
	module for irix6* systems.  Updates to the following modules:
	irixsgi, sunos5.  Fixed null pointer bug in color.c. Removed
	some useless code and definitions in display.c


Sun Nov 28 2004 - wnl (3.6alpha9)
	Replace AIX 5 module with alternate (bug 1056565).
	Fixed vulnerability in use of snprintf.

Fri Oct 22 2004 - wnl (3.6alpha8)
	Support for linux 2.6, added more stuff to memory and swap lines.
	Updated linuxthr module, which is only useful on 2.4 and earlier.
	Added some color support back in (feature request 930588), but
	still need to add it back to the per-process display.  Added
	OSF 5 support (untested).
	Fixed bug 1017951 (invalid process count argument not caught)

Tue Apr 20 2004 - wnl (3.6alpha7)
	Added 64 bit support for AIX.

Thu Apr 15 2004 - wnl (3.6alpha6)
        Included fixes for decosf1 pid size and updated module.  Also
        added osf1 to list of recognized operating systems in configure.ac.

Tue Mar 30 2004 - wnl (3.6alpha5)

old_modules/m_aux3.c
/*
 * top - a top users display for Unix
 *
 * SYNOPSIS:  a Mac running A/UX version 3.x
 *
 * DESCRIPTION:
 * This is the machine-dependent module for A/UX 3.x.
 * ==
 * Although AUX does not generally have a renice systemcall, it can be
 * implemented by tweeking kernel memory.  While such a simple hack should
 * not be difficult to get right, USE THIS FEATURE AT YOUR OWN RISK!
 * To turn on setpriority emulation, add "-DIMPLEMENT_SETPRIORITY" to
 * the CFLAGS when prompted in the configure script.
 *
 * CFLAGS: -Dclear=clear_scr -DPRIO_PROCESS=0
 *
 * LIBS:
 *
 * AUTHOR:  Richard Henderson <rth@tamu.edu>
 */


#include <stddef.h>
#include <stdio.h>
#include <string.h>
#include <errno.h>
#include <fcntl.h>
#include <a.out.h>
#include <sys/types.h>
#include <sys/signal.h>
#include <sys/param.h>
#include <sys/proc.h>
#include <sys/user.h>
#include <sys/sysinfo.h>
#include <sys/var.h>
#include <sys/swap.h>

#define FSCALE	65536.0

#include "top.h"
#include "machine.h"
#include "loadavg.h"

/*=NLIST INFO===========================================================*/

#define X_V		0
#define X_SYSINFO	1
#define X_AVENRUN	2
#define X_MAXMEM	3
#define X_FREEMEM	4
#define X_SWAPTAB	5
#define X_AVAILRMEM	6
#define X_AVAILSMEM	7

static struct nlist nlst[] = {
    {"v"},
    {"sysinfo"},
    {"avenrun"},
    {"maxmem"},
    {"freemem"},
    {"swaptab"},
    {0},		/* "availrmem" */
    {0},		/* "availsmem" */
    {0}
};

static int kmem;
static int mem;

static struct var v;

#define V_OFS		(nlst[X_V].n_value)
#define SYSINFO_OFS	(nlst[X_SYSINFO].n_value)
#define AVENRUN_OFS	(nlst[X_AVENRUN].n_value)
#define MAXMEM_OFS	(nlst[X_MAXMEM].n_value)
#define FREEMEM_OFS	(nlst[X_FREEMEM].n_value)
#define SWAPTAB_OFS	(nlst[X_SWAPTAB].n_value)
#define AVAILRMEM_OFS	(nlst[X_AVAILRMEM].n_value)
#define AVAILSMEM_OFS	(nlst[X_AVAILSMEM].n_value)

/*=SYSTEM STATE INFO====================================================*/

/* these are for calculating cpu state percentages */

static long cp_time[NCPUSTATES];
static long cp_old[NCPUSTATES];
static long cp_diff[NCPUSTATES];

/* these are for keeping track of the proc array */

struct top_proc
{
    pid_t p_pid;
    pid_t p_pgrp;
    uid_t p_uid;
    int p_pri;
    int p_nice;
    int p_size;
    int p_stat;
    int p_flag;
    int p_slot;
    time_t p_start;
    time_t p_time;
    float p_pcpu;
    float p_wcpu;
    char p_name[COMMSIZ];
};

static int hash_size;
static struct top_proc *ptable;	/* the hash table of processes */
static struct top_proc *eptable;
static struct top_proc **pactive; /* list of active structures */
static struct top_proc **nextactive; /* for iterating through the processes */
static struct proc *preal;
static struct proc *epreal;

static pid_t last_pid;
static struct timeval last_update;

/* these are for passing data back to the mach. ind. portion */

static int cpu_states[NCPUSTATES];
static int process_states[8];
static int memory_stats[6];

/* a few useful macros... */

#define blocktok(b)	((b) >> 1)
#define pagetok(pg)	((pg) << (v.v_pageshift - LOG1024))
#define HASH(x)		((x) * 1686629713UL % hash_size)

/*=STATE IDENT STRINGS==================================================*/

static char *state_abbrev[] =
{
    "", "sleep", "run", "zomb", "stop", "start", "cpu", "swap",
    NULL
};

static char *procstatenames[] =
{
    "", " sleeping, ", " running, ", " zombie, ", " stopped, ",
    " starting, ", " on cpu, ", " swapping, ",
    NULL
};

static char *cpustatenames[] =
{
    "idle", "user", "kernel", "wait", "nice",
    NULL
};

static char *memorynames[] = 
{
    "K used, ", "K free, ", "K locked   Swap: ", 
    "K used, ", "K free",
    NULL
};

static char fmt_header[] = 
  "  PID  PGRP X        PRI NICE  SIZE STATE   TIME    WCPU     CPU COMMAND";


/*======================================================================*/

int
machine_init(statics)
    struct statics *statics;
{
    /* access kernel memory */
    if (
#ifdef IMPLEMENT_SETPRIORITY
        (kmem = open("/dev/kmem", O_RDWR)) < 0 &&
#endif
	(kmem = open("/dev/kmem", O_RDONLY)) < 0)
    {
	perror("/dev/kmem");
	return -1;
    }
    if ((mem = open("/dev/mem", O_RDONLY)) < 0)
    {
	perror("/dev/mem");
	return -1;
    }

    /* get the list of symbols we want to access in the kernel */
    nlst[X_AVAILRMEM].n_nptr = "availrmem";
    nlst[X_AVAILSMEM].n_nptr = "availsmem";

    if (nlist("/unix", nlst) < 0)
    {
	fprintf(stderr, "top: nlist failed\n");
	return -1;
    }

    /* make sure they were all found */
    if (check_nlist(nlst) > 0)
	return -1;

    /* grab the kernel configuration information */
    (void)getkval(V_OFS, (char *)&v, sizeof(v), "v");

    /* allocate space for process related info */
    hash_size = v.v_proc * 3 / 2;
    ptable = (struct top_proc *)malloc(hash_size * sizeof(struct top_proc));
    pactive = (struct top_proc **)malloc(v.v_proc * sizeof(struct top_proc *));

    if (!ptable || !pactive)
    {
	fprintf(stderr, "top: can't allocate sufficient memory\n");
	return -1;
    }

    eptable = ptable + hash_size;

    {
	struct top_proc *p;
	for (p = ptable; p != eptable; ++p)
	    p->p_pid = -1;
    }

    /* fill in the statics information */
    statics->procstate_names = procstatenames;
    statics->cpustate_names = cpustatenames;
    statics->memory_names = memorynames;

    /* all done! */
    return 0;
}

static struct top_proc *
lookup_proc(id)
    pid_t id;
{
    struct top_proc *p;

    p = ptable+HASH(rp->p_pid);
    while (p->p_pid != rp->p_pid && p->p_pid != -1)
    {
	if (++p == eptable)
	    p = ptable;
    }

    return p;
}
    

static void
update_proc_table()
{
    struct proc *rp;
    struct top_proc *p;
    float timediff, alpha, beta;

    getkval((long)v.ve_proctab, (char *)preal, 
	    sizeof(struct proc)*v.v_proc, "proc array");

    /* calculate the time difference since our last proc read */
    {
	struct timeval thistime;
	gettimeofday(&thistime, 0);
	if (last_update.tv_sec)
	    timediff = ((thistime.tv_sec - last_update.tv_sec) +
		        (thistime.tv_usec - last_update.tv_usec) * 1e-6);
	else
	    timediff = 1e9;
	last_update = thistime;
    }

    /* calculate constants for the exponental average */
    if (timediff < 30.0)
    {
	alpha = 0.5 * (timediff / 30.0);
	beta = 1.0 - alpha;
    }
    else
	alpha = beta = 0.5;

    timediff *= v.v_hz;

    /* mark the hash table entries as not seen */
    for (p = ptable; p != eptable; ++p)
	p->p_stat = 0;

    for (rp = preal; rp != epreal; ++rp)
    {
	struct user u;

	if (rp->p_stat == 0)
	    continue;
	else if (rp->p_stat == SZOMB ||
		 lseek(mem, rp->p_addr, 0) < 0 ||
	         read(mem, &u, sizeof(u)) != sizeof(u))
	{
	    strcpy(u.u_comm, "???");
	    u.u_utime = u.u_stime = u.u_start = 0;
	}

	p = lookup_proc(rp->p_pid);

	p->p_pgrp = rp->p_pgrp;
	p->p_uid = rp->p_uid;
	p->p_pri = rp->p_pri - PZERO;
	p->p_nice = rp->p_nice - NZERO;
	p->p_size = pagetok(rp->p_size);
	p->p_stat = rp->p_stat;
	p->p_flag = rp->p_flag;
	if (p->p_pid != rp->p_pid)
	{
	    /* new process */
	    p->p_pid = rp->p_pid;
	    p->p_slot = rp - preal;
	    p->p_start = u.u_start;
	    p->p_time = u.u_utime + u.u_stime;
	    p->p_pcpu = p->p_time / timediff;
	    p->p_wcpu = p->p_pcpu;
	    strncpy(p->p_name, u.u_comm, sizeof(u.u_comm));
	}
	else
	{
	    time_t oldtime = p->p_time;
	    p->p_time = u.u_utime + u.u_stime;
	    p->p_pcpu = (p->p_time - oldtime) / timediff;
	    p->p_wcpu = alpha * p->p_pcpu + beta * p->p_wcpu;
	}
    }
	
    for (p = ptable; p != eptable; ++p)
	if (p->p_stat == 0)
	    p->p_pid = -1;
}

void
get_system_info(info)
    struct system_info *info;
{
    /* convert load averages */
    {
	load_avg ar[3];

	(void)getkval(AVENRUN_OFS, (char *)&ar, sizeof(ar), "avenrun");

	/* convert load averages to doubles */
	info->load_avg[0] = loaddouble(ar[0]);
	info->load_avg[1] = loaddouble(ar[1]);
	info->load_avg[2] = loaddouble(ar[2]);
    }

    /* get cpu time counts */
    {
	struct sysinfo si;

	(void)getkval(SYSINFO_OFS, (char *)&si, sizeof(si), "sysinfo");

	memcpy(cp_time, si.cpu, sizeof(cp_time));
	percentages(NCPUSTATES, cpu_states, cp_time, cp_old, cp_diff);
    }

    /* get memory usage information */
    {
	int freemem, availrmem, availsmem, maxmem;
	struct swaptab swaptab[MSFILES];
	int i, swaptot, swapfree;

        (void)getkval(MAXMEM_OFS, (char *)&maxmem, sizeof(maxmem), "maxmem");
	(void)getkval(FREEMEM_OFS, (char *)&freemem, sizeof(freemem),
		      "freemem");
	(void)getkval(AVAILRMEM_OFS, (char *)&availrmem, sizeof(availrmem),
		      "availrmem");
	(void)getkval(AVAILSMEM_OFS, (char *)&availsmem, sizeof(availsmem),
		      "availsmem");
        (void)getkval(SWAPTAB_OFS, (char *)&swaptab, sizeof(swaptab),
		      "swaptab");

	for (i = swaptot = swapfree = 0; i < MSFILES; ++i)
	    if (swaptab[i].st_dev)
	    {
		swaptot += swaptab[i].st_npgs;
		swapfree += swaptab[i].st_nfpgs;
	    }

	memory_stats[0] = pagetok(availrmem - freemem);
	memory_stats[1] = pagetok(freemem);
	memory_stats[2] = pagetok(maxmem - availrmem);
	memory_stats[3] = pagetok(swaptot - swapfree);
	memory_stats[4] = pagetok(swapfree);
    }

    update_proc_table();

    /* search proc structures for newest process id */
    {
	struct top_proc *p;
	time_t t = 0;
	pid_t id = 0;

	for (p = ptable; p != eptable; ++p)
	{
	    if (!p->p_stat)
		continue;
	    if (p->p_start > t || p->p_start == t && p->p_pid > id)
	    {
		t = p->p_start;
		id = p->p_pid;
	    }
	}

	if (id > last_pid || id < last_pid - 10000)
	    last_pid = id;

	info->last_pid = last_pid;
    }

    /* set arrays and strings */
    info->cpustates = cpu_states;
    info->memory = memory_stats;
}

caddr_t
get_process_info(si, sel, compare)
     struct system_info *si;
     struct process_select *sel;
     int (*compare)();
{
    int total_procs;
    struct top_proc *p, **a;

    /* these are copied out of sel for speed */
    int show_idle, show_system, show_uid, show_command;

    /* get a pointer to the states summary array */
    si->procstates = process_states;

    /* set up flags which define what we are going to select */
    show_idle = sel->idle;
    show_system = sel->system;
    show_uid = sel->uid != -1;
    show_command = sel->command != NULL;

    /* count up process states and get pointers to interesting procs */
    total_procs = 0;
    memset(process_states, 0, sizeof(process_states));

    for (p = ptable, a = pactive; p != eptable; ++p)
    {
	int stat = p->p_stat, flag = p->p_flag;

	if (stat == 0 || (flag & SSYS) && !show_system)
	    continue;

	total_procs++;
	process_states[stat]++;

	if (stat != SZOMB &&
	    (show_idle || stat == SRUN || stat == SIDL || stat == SONPROC ||
	     ((stat == SSLEEP || stat == SSTOP) &&
	      (flag & (SINTR | SSYS)) == 0)) &&
	     (!show_uid || p->p_uid == (uid_t)sel->uid))
	{
	    /* add it to our active list */
	    *a++ = p;
	}
    }

    /* remember active and total counts */
    si->p_total = total_procs;
    si->p_active = a - pactive;

    /* if requested, sort the "interesting" processes */
    if (compare != NULL)
	qsort(pactive, si->p_active, sizeof(struct top_proc *), compare);

    /* set up to iterate though processes */
    nextactive = pactive;

    /* don't even pretend the return value isn't bogus */
    return 0;
}


char *
format_header(uname_field)
    char *uname_field;
{
    int len = strlen(uname_field);
    if (len > 8)
	len = 8;

    memcpy(strchr(fmt_header, 'X'), uname_field, len);

    return fmt_header;
}

char *
format_next_process(handle, get_userid)
     caddr_t handle;
     char *(*get_userid)();
{
    static char fmt[128];	/* static area where result is built */
    struct top_proc *pp = *nextactive++;

    sprintf(fmt,
	    "%5d %5d %-8.8s %3d %4d %5s %-5s %6s %6.2f%% %6.2f%% %.14s",
	    pp->p_pid,
	    pp->p_pgrp,
	    (*get_userid)(pp->p_uid),
	    pp->p_pri,
	    pp->p_nice,
	    format_k(pp->p_size),
	    state_abbrev[pp->p_stat],
	    format_time((time_t)pp->p_time / v.v_hz),
	    pp->p_wcpu * 100.0,
	    pp->p_pcpu * 100.0,
	    pp->p_name);

    /* return the result */
    return (fmt);
}


/*
 * check_nlist(nlst) - checks the nlist to see if any symbols were not
 *              found.  For every symbol that was not found, a one-line
 *              message is printed to stderr.  The routine returns the
 *              number of symbols NOT found.
 */

int
check_nlist(nlst)
     register struct nlist *nlst;
{
    register int i;

    /* check to see if we got ALL the symbols we requested */
    /* this will write one line to stderr for every symbol not found */

    i = 0;
    while (nlst->n_name[0])
    {
	if (nlst->n_value == 0)
	{
	    /* this one wasn't found */
	    fprintf(stderr, "kernel: no symbol named `%s'\n", nlst->n_name);
	    i = 1;
	}
	nlst++;
    }

    return (i);
}


/*
 *  getkval(offset, ptr, size, refstr) - get a value out of the kernel.
 *      "offset" is the byte offset into the kernel for the desired value,
 *      "ptr" points to a buffer into which the value is retrieved,
 *      "size" is the size of the buffer (and the object to retrieve),
 *      "refstr" is a reference string used when printing error meessages,
 *          if "refstr" starts with a '!', then a failure on read will not
 *          be fatal (this may seem like a silly way to do things, but I
 *          really didn't want the overhead of another argument).
 *      
 */

getkval(offset, ptr, size, refstr)
     unsigned long offset;
     int *ptr;
     int size;
     char *refstr;
{
    extern int errno;
    extern char *sys_errlist[];

    if (lseek(kmem, offset, 0) < 0 || read(kmem, ptr, size) != size)
    {
	if (*refstr == '!')
	{
	    return (0);
	}
	else
	{
	    fprintf(stderr, "top: getkval for %s: %s\n",
		    refstr, sys_errlist[errno]);
	    quit(23);
	    /*NOTREACHED */
	}
    }
    return (1);
}

/* comparison routine for qsort */

/*
 *  proc_compare - comparison function for "qsort"
 *      Compares the resource consumption of two processes using five
 *      distinct keys.  The keys (in descending order of importance) are:
 *      percent cpu, cpu ticks, state, resident set size, total virtual
 *      memory usage.  The process states are ordered as follows (from least
 *      to most important):  WAIT, zombie, sleep, stop, start, run.  The
 *      array declaration below maps a process state index into a number
 *      that reflects this ordering.
 */

static unsigned char sorted_state[] =
{
    0,	/* not used             */
    3,	/* sleep                */
    6,	/* runable              */
    1,	/* zombie               */
    4,	/* stop                 */
    5,	/* start                */
    7,	/* running              */
    2,	/* swapping             */
};

proc_compare(pp1, pp2)
    struct top_proc **pp1, **pp2;
{
    struct top_proc *p1, *p2;
    int result;
    double dresult;

    /* remove one level of indirection */
    p1 = *pp1;
    p2 = *pp2;

    /* compare percent cpu */
    dresult = p2->p_pcpu - p1->p_pcpu;
    if (dresult != 0.0)
	return dresult > 0.0 ? 1 : -1;

    /* use process state to break the tie */
    if ((result = (sorted_state[p2->p_stat] -
		   sorted_state[p1->p_stat])) == 0)
    {
	/* use priority to break the tie */
	if ((result = p2->p_pri - p1->p_pri) == 0)
	{
	    /* use total memory to break the tie */
	    result = p2->p_size - p1->p_size;
	}
    }

    return result;
}

/*
 * proc_owner(pid) - returns the uid that owns process "pid", or -1 if
 *              the process does not exist.
 *              It is EXTREMLY IMPORTANT that this function work correctly.
 *              If top runs setuid root (as in SVR4), then this function
 *              is the only thing that stands in the way of a serious
 *              security problem.  It validates requests for the "kill"
 *              and "renice" commands.
 */

int
proc_owner(pid)
    int pid;
{
    struct top_proc *p;

    for (p = ptable; p != eptable; ++p)
	if (p->p_pid == pid)
	    return p->p_uid;

    return -1;
}

/* 
 * setpriority(int which, pid_t pid, int val)
 * This system does not have this system call -- fake it
 */

int
setpriority(which, pid, val)
    int which, pid, val;
{
#ifndef IMPLEMENT_SETPRIORITY
    errno = ENOSYS;
    return -1;
#else
    struct top_proc *p;
    struct proc proc;
    int uid;

    /* sanity check arguments */
    val += NZERO;
    if (val < 0)
	val = 0;
    else if (val > 39)
	val = 39;

    p = lookup_proc(pid);
    if (p->p_pid == -1)
    {
	errno = ESRCH;
	return -1;
    }

    getkval((long)v.ve_proctab+p->p_slot*sizeof(proc),
	    (char *)&proc, sizeof(proc), "proc array");

    if (proc.p_stat == 0 || proc.p_pid != pid)
    {
	errno = ESRCH;
	return -1;
    }

    /* make sure we don't allow nasty people to do nasty things */
    uid = getuid();
    if (uid != 0)
    {
	if (uid != proc.p_uid || val < proc.p_nice)
	{
	    errno = EACCES;
	    return -1;
	}
    }

    /* renice */
    proc.p_nice = val;
    if (lseek(kmem, (v.ve_proctab + p->p_slot*sizeof(proc) +
		     offsetof(struct proc, p_nice), 0) < 0 ||
	write(kmem, &rp->p_nice, sizeof(rp->p_nice)) != sizeof(rp->p_nice))
    {
	return -1;
    }

    return 0;
#endif
}
